sql-parser/src/main/java/io/crate/sql/Literals.java

```diff
@@ -24,14 +24,163 @@
 
 import com.google.common.annotations.VisibleForTesting;
 
+import java.util.Arrays;
+import java.util.function.Predicate;
+
 public class Literals {
 
+    private static final String EPSILON = "E";
+    private static final String ESCAPED_UNICODE_ERROR = "Invalid Unicode escape (must be \\uXXXX or \\UXXXXXXXX)";
+
     public static String quoteStringLiteral(String literal) {
         return "'" + escapeStringLiteral(literal) + "'";
     }
 
+    public static String escapeAndQuoteStringLiteral(String literal) {
+        return EPSILON + Literals.quoteStringLiteral(literal);
+    }
+
     @VisibleForTesting
     static String escapeStringLiteral(String literal) {
         return literal.replace("'", "''");
     }
+
+    public static String replaceEscapedChars(String input) {
+        if (input.length() <= 1) {
+            return input;
+        }
+
+        char[] inputCharArray = input.toCharArray();
+        StringBuilder builder = new StringBuilder(input.length());
+        int prev = 0;
+        int current = 0;
+
+        boolean firstToken = true;
+        char[] tokenCharArray;
+        int tokenIdx, tokenLen;
+
+        while (current < inputCharArray.length) {
+            while (current < inputCharArray.length && '\\' != inputCharArray[current]) {
+                current++;
+            }
+            tokenCharArray = Arrays.copyOfRange(inputCharArray, prev, current);
```

Comment:
How can I remove this copy operation here? I just need a reference for this sub-array.

Reply:
you can use `builder.append(inputCharArray, prev, current)` instead of creating a new array slice.